cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java

```diff
@@ -42,11 +50,16 @@
   private String programName;
   private Runnable programRunner;
   private SchedulableProgramType programType;
+  private Map<String, Dataset> datasets;
+  private TransactionSystemClient txClient;
 
-  public ProgramWorkflowAction(String name, String programName, SchedulableProgramType programType) {
+  public ProgramWorkflowAction(String name, String programName, SchedulableProgramType programType,
+                               @Nullable TransactionSystemClient txClient) {
```

Comment:
WorkflowDriver calls the `action.run` in order to execute the action. For programs (like MapReduce/Spark), it is the run method on the ProgramWorkflowAction which eventually runs the program, however for custom actions its the user defined `run` method that would get called. We wanted a wrapper on the custom action's `run` method as well, so that inside the wrapper's `run` method we can execute the custom action's `run` method within the transaction. 

For this reason, the ProgramWorkflowAction also used for running the custom actions run method.